yamux/src/control.rs

```diff
@@ -0,0 +1,156 @@
+use futures::{
+    channel::{mpsc, oneshot},
+    sink::SinkExt,
+    FutureExt,
+};
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use crate::{error::Error, stream::StreamHandle};
+
+pub(crate) enum Command {
+    OpenStream(oneshot::Sender<Result<StreamHandle, Error>>),
+    Shutdown(oneshot::Sender<()>),
+}
+
+/// A session control is used to open the stream or close the session
+pub struct Control {
+    sender: mpsc::Sender<Command>,
+    pending_open: Option<oneshot::Receiver<Result<StreamHandle, Error>>>,
+    pending_close: Option<oneshot::Receiver<()>>,
+}
+
+impl Control {
+    pub(crate) fn new(sender: mpsc::Sender<Command>) -> Self {
+        Control {
+            sender,
+            pending_open: None,
+            pending_close: None,
+        }
+    }
+
+    /// Open a new stream to remote session
+    pub async fn open_stream(&mut self) -> Result<StreamHandle, Error> {
+        let (tx, rx) = oneshot::channel();
+        self.sender
+            .send(Command::OpenStream(tx))
+            .await
+            .map_err(|_| Error::SessionShutdown)?;
+        rx.await.map_err(|_| Error::SessionShutdown)?
+    }
+
+    /// shutdown is used to close the session and all streams.
+    pub async fn close(&mut self) {
+        if self.sender.is_closed() {
+            return;
+        }
+        let (tx, rx) = oneshot::channel();
+        let _ignore = self.sender.send(Command::Shutdown(tx)).await;
+        let _ignore = rx.await;
+    }
+
+    /// Poll base open stream
+    pub fn poll_open_stream(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context,
+    ) -> Poll<Result<StreamHandle, Error>> {
```

Comment:
Where is this function been used?

Reply:
I don‘t know，maybe someone like poll base function？ this is a public funtion